### PR TITLE
Add attach method to obtain unique id

### LINKED ISSUE #3831: No way to filter the actual opening overlay - the show method returns undefined id at that moment.
### Description  
onOpening event is fired for all overlays on the page and there is no way to filter the actual opening overlay because the show method returns undefined id at that moment.

 * igniteui-angular version: 7.2
 * browser: NA

### Steps to reproduce  

1. Add two overlays on a page and subscribe to onOpening event
2. onOpening event is called twice  

### Result  
onOpening event is fired for all existing overlays on the page

### Expected result  
Have a way to identify the actual opening overlay

### PR DESCRIPTION
**Pull request by wnvko**
`attach` method added to `igxOverlay`. Developer may call this method to obtain unique Id for the overlay where provided component should be shown. Later developer may call `show(id)` method to show the component.

Closes #3831

### Additional information (check all that apply):
 - [x] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code
 - [ ] This PR includes API docs for newly added methods/properties
 - [x] This PR includes `feature/README.MD` updates for the feature docs
 - [x] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts [hunk 1]
```diff
@@ -184,11 +190,16 @@ export class IgxOverlayService implements OnDestroy {
             this.setupModalWrapper(info);
         }
 
-        if (info.settings.positionStrategy.settings.openAnimation) {
-            this.playOpenAnimation(info);
-        } else {
-            this.onOpened.emit({ id: info.id, componentRef: info.componentRef });
-        }
+        //  show overlay in set time out. This allow us to return the id before overlay appears
+        setTimeout(() => {
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: Slight issue - `onOpening` still emits way in advance, so you can't actually intercept it with the returned id.

Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts [hunk 2]
```diff
@@ -143,14 +149,69 @@ export class IgxOverlayService implements OnDestroy {
         settings = Object.assign({}, this._defaultSettings, settings);
         info.settings = settings;
 
-        const eventArgs = { id, componentRef: info.componentRef, cancel: false };
+        //  show overlay in set time out. This allow us to return the id before overlay appears
+        setTimeout(() => {
```
Comment (kdinev) [inline projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: Unfortunately, this is a workaround. Please look for a way to do this without the `setTimeout`. I wanna see no evil and setTimeout === evil.

Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: we add a new method - register (we may choose and better name, e.g. attach). This method will prepare the overlay and will return the ID. Then developer can show the overlay via show(id, settings) call. We are deprecating show(component, settings) method.

File: CHANGELOG.md [hunk 3]
```diff
@@ -80,6 +80,9 @@ All notable changes for each version of this project will be documented in this
 - `Column Hiding UI`
     - **Behavioral Change** - The UI now hides the columns whose `disableHiding` property is set to true instead of simply disabling them.
 - `igxButton` - **New Button Style** - Include [outlined](https://material.io/design/components/buttons.html#outlined-button) button style to support the latest material spec.
+- `igxOverlay`:
+    - `igxOverlay.register(component, settings?)` method added. Use this method to obtain Id for the component you need to show in overlay. Then call `igxOverlay.show(id, settings?)` method to show the component in overlay.
+    - `igxOverlay.show(component, settings)` is **depricated**. Use `igxOverlay.register(component, settings?)` method to obtain Id, and then call `igxOverlay.show(id, settings)` method to show a component in the overlay.
```
Comment (bazal4o) [inline CHANGELOG.md]: obtain an Id

File: CHANGELOG.md [hunk 4]
```diff
@@ -80,6 +80,9 @@ All notable changes for each version of this project will be documented in this
 - `Column Hiding UI`
     - **Behavioral Change** - The UI now hides the columns whose `disableHiding` property is set to true instead of simply disabling them.
 - `igxButton` - **New Button Style** - Include [outlined](https://material.io/design/components/buttons.html#outlined-button) button style to support the latest material spec.
+- `igxOverlay`:
+    - `igxOverlay.register(component, settings?)` method added. Use this method to obtain Id for the component you need to show in overlay. Then call `igxOverlay.show(id, settings?)` method to show the component in overlay.
```
Comment (bazal4o) [inline CHANGELOG.md]: obtain an unique Id of the created overlay where the provided component will be shown

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts [hunk 5]
```diff
@@ -101,6 +107,32 @@ export class IgxOverlayService implements OnDestroy {
...
 
+    /**
+     * Generates Id. Provide this Id when call `show(id, settings?)` method
+     * @param component ElementRef or Component Type to show in overlay
+     * @param settings Display settings for the overlay, such as positioning and scroll/close behavior.
+     * @returns Id of the created overlay. Valid until `onClosed` is emitted.
+     */
+    register(component: ElementRef | Type<{}>, settings?: OverlaySettings): string {
+        let info: OverlayInfo;
+        const id = (this._componentId++).toString();
```
Comment (bazal4o) [inline projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: Move the id generation after the return statement. So the componentId doesn't get increased if info is not correct

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts [hunk 6]
```diff
@@ -187,20 +277,10 @@ export class IgxOverlayService implements OnDestroy {
         if (info.settings.positionStrategy.settings.openAnimation) {
             this.playOpenAnimation(info);
         } else {
+            //  to eliminate flickering show the element just before onOpened fire
+            info.elementRef.nativeElement.parentElement.visibility = null;
```
Comment (bazal4o) [inline projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: Do we need to set it to null or delete will work too?

File: CHANGELOG.md (hunk 3 above)
Comment (wnvko) [reply CHANGELOG.md]: done

File: CHANGELOG.md (hunk 4 above)
Comment (wnvko) [reply CHANGELOG.md]: done

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts (hunk 5 above)
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts (hunk 6 above)
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: delete will not work here, as per my tests :)

**Comment by kdinev** [comment]
Change the title of the PR, if it doesn't include a `setTimeout` anymore.

**Pull request merged**